File: src/jskills/factorgraphs/Factor.java
package jskills.factorgraphs;

import static jskills.Guard.argumentIsValidIndex;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class Factor<TValue> {

    private final List<Message<TValue>> messages = new ArrayList<Message<TValue>>();

    private final Map<Message<TValue>, Variable<TValue>> messageToVariableBinding =
        new HashMap<Message<TValue>, Variable<TValue>>();

    private final String name;
    private final List<Variable<TValue>> variables = new ArrayList<Variable<TValue>>();

    protected Factor(String name) { this.name = "Factor[" + name + "]"; }

    /** Returns the log-normalization constant of that factor **/
    public abstract double getLogNormalization();

    /** Returns the number of messages that the factor has **/
    public int getNumberOfMessages() { return messages.size(); }

    protected Collection<Variable<TValue>> getVariables() {
        return Collections.unmodifiableCollection(variables); 
    }

    protected Collection<Message<TValue>> getMessages() {
        return Collections.unmodifiableCollection(messages);
    }

    /** Update the message and marginal of the i-th variable that the factor is connected to **/
    public double updateMessage(int messageIndex) {
        argumentIsValidIndex(messageIndex, messages.size(), "messageIndex");
        return UpdateMessage(messages.get(messageIndex), messageToVariableBinding.get(messages.get(messageIndex)));
    }

    protected abstract double UpdateMessage(Message<TValue> message, Variable<TValue> variable);

    /** Resets the marginal of the variables a factor is connected to **/
    public void ResetMarginals() {
        for(Variable<TValue> variable : messageToVariableBinding.values())
            variable.resetToPrior();
    }

    /**
     * Sends the ith message to the marginal and returns the log-normalization
     * constant
     **/
    public double SendMessage(int messageIndex) {
        argumentIsValidIndex(messageIndex, messages.size(), "messageIndex");

        Message<TValue> message = messages.get(messageIndex);
        Variable<TValue> variable = messageToVariableBinding.get(message);
        return SendMessage(message, variable);
    }

    protected abstract double SendMessage(Message<TValue> message, Variable<TValue> variable);

    public abstract Message<TValue> CreateVariableToMessageBinding(Variable<TValue> variable);

    protected Message<TValue> CreateVariableToMessageBinding(Variable<TValue> variable, Message<TValue> message) {
        messages.add(message);
        messageToVariableBinding.put(message, variable);
        variables.add(variable);

        return message;
    }

    @Override
    public String toString() { return name != null ? name : super.toString(); }
}

File: src/main/scala/jskills/factorgraphs/Factor.scala
package jskills.factorgraphs

import jskills.Guard._
import collection.mutable.Map
import scala.collection.mutable.Seq
import scala.collection.mutable.ListBuffer

abstract class Factor[T](name: String, val messages: ListBuffer[Message[T]] = ListBuffer.empty[Message[T]]) {
  val messageToVariableBinding = Map.empty[Message[T], Variable[T]]
  val variables = ListBuffer.empty[Variable[T]]

  /** Returns the log-normalization constant of that factor **/
  def getLogNormalization(): Double

  /** Returns the number of messages that the factor has **/
  def getNumberOfMessages(): Int = messages.size

  /** Update the message and marginal of the i-th variable that the factor is connected to **/
  def updateMessage(messageIndex: Int): Double = {
    argumentIsValidIndex(messageIndex, messages.size, "messageIndex")
     updateMessage(messages(messageIndex), messageToVariableBinding(messages(messageIndex)))
  }

  protected def updateMessage(message: Message[T], variable: Variable[T]): Double =
    throw new UnsupportedOperationException()

  /** Resets the marginal of the variables a factor is connected to **/
  def resetMarginals() { messageToVariableBinding.values foreach (_.resetToPrior) }

  /**
   * Sends the ith message to the marginal and s the log-normalization
   * constant
   */
  def sendMessage(messageIndex: Int): Double = {
    argumentIsValidIndex(messageIndex, messages.size, "messageIndex")
    val message = messages(messageIndex)
    val variable = messageToVariableBinding(message)
     sendMessage(message, variable)
  }

  protected def sendMessage(message: Message[T], variable: Variable[T]): Double

  def createVariableToMessageBinding(variable: Variable[T]): Message[T]

  protected def createVariableToMessageBinding(variable: Variable[T], message: Message[T]): Message[T] = {
    messages += message
    messageToVariableBinding.put(message, variable)
    variables += variable
     message
  }

  override def toString() = if (name != null) name else super.toString()
}